### PR TITLE
OCPBUGS-36404: Filter CSRs by signerName

### PR DESCRIPTION
**Pull request by RadekManak**
Updates to CSR filtering to utilize signerName field using the well known certificatesv1.KubeAPIServerClientKubeletSignerName and certificatesv1.KubeletServingSignerName.

Also updates the CSR query to use node selectors with signerName to be more efficient on cluster with many irrelevant CSRs

### CODE REVIEW

**Review by elmiko — COMMENTED** [review]
i think this makes sense to me
/lgtm

**Comment by elmiko** [comment]
hmm, seems like the unit tests might be broken?

/lgtm cancel

**Comment by RadekManak** [comment]
/hold
This is failing because of following error
```2024-10-22T11:33:58.542278650Z E1022 11:33:58.542266       1 controller.go:316] "Reconciler error" err="system:openshift:openshift-monitoring-8lh5v: failed to list CSRs: failed to get CSRs: Index with name field:spec.signerName does not exist" controller="certificatesigningrequest" controllerGroup="certificates.k8s.io" controllerKind="CertificateSigningRequest" CertificateSigningRequest="system:openshift:openshift-monitoring-8lh5v" namespace="" name="system:openshift:openshift-monitoring-8lh5v" reconcileID="f266ce9b-493a-412f-a1c5-0c54eaab8140"```

I remember reading docs that spec.signerName should work as selector on certificatesigningrequest. But maybe I am doing something wrong. I'll take a look when I have time. Worst case, we list all and like we are doing now and filter during reconcile.

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -115,12 +115,42 @@ func caConfigMapFilter(obj runtime.Object, new runtime.Object) bool {
 		cmData != cmDataNew
 }
 
+func listNodeCSRs(ctx context.Context, nodeClient client.Client) ([]certificatesv1.CertificateSigningRequest, error) {
+	csrList := &certificatesv1.CertificateSigningRequestList{}
+	csrs := []certificatesv1.CertificateSigningRequest{}
+
+	if err := nodeClient.List(ctx, csrList, &client.ListOptions{FieldSelector: fields.OneTermEqualSelector(signerNameField, certificatesv1.KubeAPIServerClientKubeletSignerName)}); err != nil {
```
[the remark this answers is not in the document]
Comment (RadekManak) [reply pkg/controller/controller.go]: I've taken the name from the client's name that is passed to it. Agree it should be generic in the function signature.

I've also renamed MachineClient -> ManagementClient to match uncachedManagementClient and NodeClient -> WorkloadClient to match uncachedManagementClient. I think this naming better conveys the importance of using the right client. The previous naming communicates different uses, but the new naming explains that the clients could be talking to different clusters.

File: pkg/controller/controller.go [hunk 2]
```diff
@@ -115,12 +115,42 @@ func caConfigMapFilter(obj runtime.Object, new runtime.Object) bool {
...
+		return nil, fmt.Errorf("failed to get CSRs: %w", err)
+	}
+	csrs = append(csrs, csrList.Items...)
+
+	return csrs, nil
+}
+
 func (m *CertificateApprover) Reconcile(ctx context.Context, req ctrl.Request) (reconcile.Result, error) {
-	csrs := &certificatesv1.CertificateSigningRequestList{}
+	// Ensure we filter unwanted CSRs on controller startup. These reconciles don't come from events and are not subject to the event filter.
```
[the remark this answers is not in the document]
Comment (RadekManak) [reply pkg/controller/controller.go:166]: I've seen CSRs getting past the filter on controller startup. I've added some extra logging. Let's wait and see what it does now.

File: pkg/controller/csr_check_test.go [hunk 3]
```diff
@@ -1966,20 +1967,27 @@ func TestRecentlyPendingNodeBootstrapperCSRs(t *testing.T) {
...
+			SignerName: certificatesv1.KubeAPIServerClientKubeletSignerName,
 		},
 	}
-	pendingCSR := certificatesv1.CertificateSigningRequest{}
 	multusCSR := certificatesv1.CertificateSigningRequest{
 		Spec: certificatesv1.CertificateSigningRequestSpec{
-			Username: nodeUserPrefix + "clustername-abcde-master-us-west-1a-0",
-			Request:  []byte(multusCSRPEM),
+			Username:   nodeUserPrefix + "clustername-abcde-master-us-west-1a-0",
+			SignerName: certificatesv1.KubeAPIServerClientSignerName,
```
[the remark this answers is not in the document]
Comment (RadekManak) [reply pkg/controller/csr_check_test.go]: The signername is not one we reconcile, and it correctly does not increase the csr count. I could not find an example of maltus CSR. I've made the test even more generic. Do you have an example CSR for maltus to recreate the test?

**Comment by RadekManak** [comment]
I've updated the filter to only allow kubernetes.io/kubelet-serving is they have system:nodes group and kubernetes.io/kube-apiserver-client-kubelet when they are created by nodeBootstrapper. 

This should correctly filter out all csrs we don't want to reconcile.

File: pkg/controller/csr_check_test.go [hunk 4]
```diff
@@ -1966,20 +1967,40 @@ func TestRecentlyPendingNodeBootstrapperCSRs(t *testing.T) {
...
+		Spec: certificatesv1.CertificateSigningRequestSpec{
+			Username:   "multus-pod",
+			SignerName: certificatesv1.KubeAPIServerClientKubeletSignerName,
+			Request:    []byte(multusCSRPEM),
+		},
+	}
+	pendingOtherApiserverClient := certificatesv1.CertificateSigningRequest{
+		Spec: certificatesv1.CertificateSigningRequestSpec{
+			Username:   "otherUsername",
+			SignerName: certificatesv1.KubeAPIServerClientSignerName,
```
Comment (JoelSpeed) [inline pkg/controller/csr_check_test.go:1997]: Lets make this something obviously different to the others (it looks fairly similar for now) and then I think this PR is good to go

**Review by JoelSpeed — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**